File: API/PetCenterModels/DataTransferObjects/Individual.cs/IndividualResponseDTO.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using PetCenterModels.DBTables;


namespace PetCenterModels.DataTransferObjects
{
    public class MedicalEntrySubDTO : IBaseResponseDTO<MedicalRecordEntry, MedicalEntrySubDTO>
    {
        public Guid? Id {get; set;}
        public List<NoteSubDTO>? Notes {get; set;}

        public DateTime DatePerformed {get; set;} = DateTime.UtcNow;

        public Guid ProcedureId {get; set;} = Guid.Empty;

        public Guid AnimalId{get; set;} = Guid.Empty;

        public static MedicalEntrySubDTO? FromEntity(MedicalRecordEntry? ent)
        {
            if(ent==null){return null;}
            return new MedicalEntrySubDTO
            {
                Id=ent.Id,
                DatePerformed=ent.DatePerformed,
                ProcedureId=ent.ProcedureId,
                AnimalId=ent.AnimalId
            };
        }
    }
    public class IndividualResponseDTO : IBaseResponseDTO<Individual,IndividualResponseDTO>
    {        

        public Guid? Id {get; set;}

        public Guid Identity {get; set;} = Guid.Empty;
        public string Name {get; set;} = string.Empty;

        public Guid BreedId {get; set;} = Guid.Empty;

        public bool Sex {get; set;}

        public DateTime BirthDate {get; set;} = DateTime.UtcNow;

        public List<NoteSubDTO>? Notes {get; set;}

        public List<MedicalEntrySubDTO> MedicalRecord {get; set;} = new();

        public static IndividualResponseDTO? FromEntity(Individual? ind)
        {
            if(ind==null){return null;}

            return new IndividualResponseDTO
            {
                Id = ind.Id,
                Identity=ind.AnimalIdentity,
                Name = ind.Name,
                BreedId=ind.BreedId,
                Sex=ind.Sex,
                BirthDate=ind.BirthDate,
                MedicalRecord = ind.MedicalRecord.Select(m=>MedicalEntrySubDTO.FromEntity(m)!).ToList()
            };
        }
    }
}

File: Frontend/pet_center_app/lib/models/data_transfer/individual/individual_response_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pet_center_app/models/data_transfer/note_sub_dto.dart';
part 'individual_response_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class MedicalEntrySubDTO {
  String? id;
  List<NoteSubDTO>? notes;
  String currentVersion;
  DateTime datePerformed;
  String procedureId;
  String animalId;

  MedicalEntrySubDTO({
    this.id,
    this.notes,
    this.currentVersion = '',
    DateTime? datePerformed,
    this.procedureId = '',
    this.animalId = '',
  }) : datePerformed = datePerformed ?? DateTime.now().toUtc();

  factory MedicalEntrySubDTO.fromJson(Map<String, dynamic> json) =>
      _$MedicalEntrySubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$MedicalEntrySubDTOToJson(this);

  MedicalEntrySubDTO copy() => MedicalEntrySubDTO(
    id: id,
    notes: notes?.map((n) => n.copy()).toList(),
    currentVersion: currentVersion,
    datePerformed: datePerformed,
    procedureId: procedureId,
    animalId: animalId,
  );
}

@JsonSerializable(explicitToJson: true)
class IndividualResponseDTO {
  String? id;
  String currentVersion;
  String identity;
  String name;
  String breedId;
  bool sex;
  DateTime birthDate;
  List<NoteSubDTO>? notes;
  List<MedicalEntrySubDTO> medicalRecord;

  IndividualResponseDTO({
    this.id,
    this.currentVersion = '',
    this.identity = '',
    this.name = '',
    this.breedId = '',
    this.sex = false,
    DateTime? birthDate,
    this.notes,
    List<MedicalEntrySubDTO>? medicalRecord,
  }) : birthDate = birthDate ?? DateTime.now().toUtc(),
       medicalRecord = medicalRecord ?? [];

  factory IndividualResponseDTO.fromJson(Map<String, dynamic> json) =>
      _$IndividualResponseDTOFromJson(json);

  Map<String, dynamic> toJson() => _$IndividualResponseDTOToJson(this);

  IndividualResponseDTO copy() => IndividualResponseDTO(
    id: id,
    currentVersion: currentVersion,
    identity: identity,
    name: name,
    breedId: breedId,
    sex: sex,
    birthDate: birthDate,
    notes: notes?.map((n) => n.copy()).toList(),
    medicalRecord: medicalRecord.map((m) => m.copy()).toList(),
  );
}
